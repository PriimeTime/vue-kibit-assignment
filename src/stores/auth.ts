import { defineStore } from "pinia";
import { ref, type Ref } from "vue";

export const useAuthStore = defineStore("auth", {
  state: () => ({
    user: ref(null) as Ref<User | null>,
  }),
  actions: {
    setUser(user: User | null) {
      this.user = user;
    },
    logout() {
      this.user = null;
    },
  },
});
