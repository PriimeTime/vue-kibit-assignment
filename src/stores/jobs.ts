import { ref, computed, type Ref } from "vue";
import { defineStore } from "pinia";
import type { S } from "vitest/dist/chunks/config.Cy0C388Z.js";

export const useJobsStore = defineStore("jobs", {
  state: () => ({
    loading: false, // TODO add this to a global loading hook ad create a global loading component
    searchTerm: ref("") as Ref<string>,
    jobs: [] as Job[],
  }),
  getters: {
    filteredJobs: (state) => (searchString: string) => {
      if (!searchString) {
        return state.jobs;
      }

      const lowercasedSearch = searchString.toLowerCase();
      return state.jobs.filter(
        (job) =>
          job.title.toLowerCase().includes(lowercasedSearch) ||
          job.description.toLowerCase().includes(lowercasedSearch) ||
          job.tags.some((tag) => tag.toLowerCase().includes(lowercasedSearch))
      );
    },
  },
});
