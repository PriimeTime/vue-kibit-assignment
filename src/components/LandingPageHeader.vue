<script setup lang="ts">
import BaseButton from "@/components/BaseButton.vue";
import BaseInput from "@/components/BaseInput.vue";
import { useJobsStore } from "@/stores/jobs";
import { ref, watch } from "vue";
import BaseCard from "@/components/BaseCard.vue";
import HeaderText from "@/components/HeaderText.vue";

const jobStore = useJobsStore();

const searchInputValue = ref("");

const search = () => {
  jobStore.searchTerm = searchInputValue.value;
};

const handleKeyDown = (event: KeyboardEvent) => {
  if (event.key === "Enter") {
    search();
  }
};

// Reset search results when search input is empty
watch(searchInputValue, () => {
  if (searchInputValue.value === "") {
    search();
  }
});
</script>

<template>
  <section class="w-full flex justify-center">
    <BaseCard>
      <HeaderText>Find the job of your dreams</HeaderText>
      <div class="w-full m-2 flex justify-between">
        <BaseInput
          @keydown="handleKeyDown"
          class="w-full"
          placeholder="Search for jobs"
          v-model="searchInputValue"
        ></BaseInput>
        <BaseButton @click="search" class="mx-2">Search</BaseButton>
      </div>
    </BaseCard>
  </section>
</template>
