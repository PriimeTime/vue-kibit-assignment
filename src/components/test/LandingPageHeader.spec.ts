import { describe, it, expect, vi } from "vitest";
import { mount } from "@vue/test-utils";
import LandingPageHeader from "@/components/LandingPageHeader.vue";
import { createTestingPinia } from "@pinia/testing";
import { useJobsStore } from "@/stores/jobs";

const testConfig = {
  global: {
    plugins: [
      createTestingPinia({
        createSpy: vi.fn,
      }),
    ],
  },
};

describe("LandingPageHeader", () => {
  it("renders without crashing", () => {
    const wrapper = mount(LandingPageHeader, testConfig);
    expect(wrapper.exists()).toBe(true);
  });

  it("updates store's searchTerm when search button is clicked", async () => {
    const wrapper = mount(LandingPageHeader, testConfig);

    const searchInput = wrapper.find("input");
    const searchButton = wrapper.find("button");

    await searchInput.setValue("developer");
    await searchButton.trigger("click");

    const jobStore = useJobsStore();
    expect(jobStore.searchTerm).toBe("developer");
  });

  it("triggers search on pressing Enter key", async () => {
    const wrapper = mount(LandingPageHeader, testConfig);

    const searchInput = wrapper.find("input");

    await searchInput.setValue("frontend");
    await searchInput.trigger("keydown", { key: "Enter" });

    const jobStore = useJobsStore();
    expect(jobStore.searchTerm).toBe("frontend");
  });

  it("resets search results when input is cleared", async () => {
    const wrapper = mount(LandingPageHeader, testConfig);

    const searchInput = wrapper.find("input");
    const jobStore = useJobsStore();

    await searchInput.setValue("designer");
    await searchInput.setValue("");

    expect(jobStore.searchTerm).toBe("");
  });

  it("matches snapshot", () => {
    const wrapper = mount(LandingPageHeader, testConfig);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
