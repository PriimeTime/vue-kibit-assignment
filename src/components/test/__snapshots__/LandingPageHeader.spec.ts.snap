// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LandingPageHeader > matches snapshot 1`] = `
"<section class="w-full flex justify-center">
  <div class="max-w-[40rem] rounded-xl md:shadow-xl p-10 w-full">
    <div class="text-3xl font-bold">Find the job of your dreams!</div>
    <div class="w-full m-2 flex justify-between">
      <div class="w-full"><input class="border-teal-900 border-2 rounded-lg p-2 w-full focus:outline-none" placeholder="Search for jobs" value=""></div><button class="p-2 min-w-20 rounded-md text-white border-teal-900 border-2 bg-teal-900 hover:text-teal-900 hover:bg-white flex justify-center items-center whitespace-nowrap mx-2">Search</button>
    </div>
  </div>
</section>"
`;
